### ISSUE TITLE
Add new authentication method (OAuth2) for email accounts

### ISSUE THREAD

**Issue by RomanovM** [nopSolutions/nopCommerce#7031]
nopCommerce version: 4.60

Let's add a new field editor (select list) on the "Edit email account details" page with the Name - "Authentication Method" and the following values:
- No authentication (no additional fields)
- Login/Password (Login and Password fields)
- OAuth2 (ClientId and ClientSecret fields)

For OAuth2 authentication we also should implement the appropriate feature in SMTP client we use.

**Comment by SelectSystemsInternational** [comment]
If you mean OAuth2 for Outlook - it is a bit of a complicated setup in Microsoft 365 itself - firstly to create an App with the required settings and access privs.  Then I modified SendEmailAsync 
It takes three settings - as a short cut I currently I use the Password in the format: AppId;TenantId;ClientSecret

```
        public virtual async Task SendEmailAsync(EmailAccount emailAccount, string subject, string body,
            string fromAddress, string fromName, string toAddress, string toName,
            string replyTo = null, string replyToName = null,
            IEnumerable<string> bcc = null, IEnumerable<string> cc = null,
            string attachmentFilePath = null, string attachmentFileName = null,
            int attachedDownloadId = 0, IDictionary<string, string> headers = null)
        {
            var message = new MimeMessage();

            //Existing Code...

            message.Body = multipart;

            if (emailAccount is null)
            {
                emailAccount = await _emailAccountService.GetEmailAccountByIdAsync(_emailAccountSettings.DefaultEmailAccountId)
                ?? throw new NopException("Email account could not be loaded");
            }

            var client = new MailKit.Net.Smtp.SmtpClient();

            try
            {
                string appId = string.Empty; 
                string tenantId = string.Empty; 
                string clientSecret = string.Empty; 

                var credentials = emailAccount.Password.Split(";");
                if (credentials.Length != 3)
                {
                    // error
                }
                else
                {
                    appId = credentials[0];
                    tenantId = credentials[1];
                    clientSecret = credentials[2];
                }

                var confidentialClientApplication = ConfidentialClientApplicationBuilder.Create(appId)
                    .WithAuthority($"https://login.microsoftonline.com/{tenantId}/v2.0")
                    .WithClientSecret(clientSecret)
                    .Build();

                var scopes = new string[] { "https://outlook.office365.com/.default" };

                SaslMechanismOAuth2 oauth2 = null;
                if (emailAccount.Host.Contains("outlook.office365.com"))
                {
                    var authToken = await confidentialClientApplication.AcquireTokenForClient(scopes).ExecuteAsync();
                    oauth2 = new SaslMechanismOAuth2(emailAccount.Email, authToken.AccessToken);
                    client.Connect(emailAccount.Host);
                }
                else
                {
                    client.ServerCertificateValidationCallback = _smtpBuilder.ValidateServerCertificate;

                    await client.ConnectAsync(emailAccount.Host, emailAccount.Port,
                        emailAccount.EnableSsl ? SecureSocketOptions.SslOnConnect : SecureSocketOptions.StartTlsWhenAvailable);
                }

                if (emailAccount.Host.Contains("outlook.office365.com"))
                {
                    client.Authenticate(oauth2);
                    client.Send(message);
                }
                else if (emailAccount.UseDefaultCredentials)
                {
                    await client.AuthenticateAsync(CredentialCache.DefaultNetworkCredentials);
                }
                else if (!string.IsNullOrWhiteSpace(emailAccount.Username))
                {
                    await client.AuthenticateAsync(new NetworkCredential(emailAccount.Username, emailAccount.Password));
                }
            }
```
For more info See [https://github.com/jstedfast/MailKit/blob/master/ExchangeOAuth2.md#authenticating-a-web-service-with-oauth2](url)

**Comment by RomanovM** [comment]
@SelectSystemsInternational Thanks for the information. We're initially going to add OAuth2 support for Gmail accounts, but with your help, we probably add it for Microsoft accounts as well.

**Comment by SelectSystemsInternational** [comment]
Hi Anokhin,

 

Everything worked fine and the email was sent except I had to change the Connect

 

The default call gave an exception Message = "The operation has timed out."

 

I checked the port was correct and played with a few combinations for options

but seems the only one that worked was use the Host only with the rest default values - Port needs to be 0 for some reason ?

 

This worked 

 

            if (emailAccount.EmailAuthenticationMethod != EmailAuthenticationMethod.MicrosoftOAuth2)

                await client.ConnectAsync(

                          emailAccount.Host,

                          emailAccount.Port,

                          emailAccount.EnableSsl ? SecureSocketOptions.SslOnConnect : SecureSocketOptions.StartTlsWhenAvailable);

            else

                await client.ConnectAsync(emailAccount.Host);

 

 

PS Also check the UI Error – when no Send to email specified makes a UI issue 

 

Let me know if you want to check anything else 

 

Thanks

Andrew

**Comment by SelectSystemsInternational** [comment]
Hi Anokhin,

Everything worked fine and the email was sent except I had to change the Connect
The default call gave an exception Message = **"The operation has timed out."**

I checked the port was correct and played with a few combinations for options
but seems the only one that worked was use the Host only with the rest default values
- Port needs to be 0 for some reason ?
 
This worked 

            if (emailAccount.EmailAuthenticationMethod != EmailAuthenticationMethod.MicrosoftOAuth2)
                await client.ConnectAsync(
                          emailAccount.Host,
                          emailAccount.Port,
                          emailAccount.EnableSsl ? SecureSocketOptions.SslOnConnect : SecureSocketOptions.StartTlsWhenAvailable);
            else
                await client.ConnectAsync(emailAccount.Host);

PS Also check the UI Error – when no Send to email specified makes a UI issue 
![UI Error](https://github.com/nopSolutions/nopCommerce/assets/34350154/df38ca8a-5722-4165-982a-17f86119a934)

**Comment by exileDev** [comment]
> I checked the port was correct i.e. 465 and played with a few combinations for options
but seems the only values that worked was to use the Host only with the rest default values

@SelectSystemsInternational Thank you! According to MailKit sources, it should work with Port 25 and unchecked SSL

**Comment by SelectSystemsInternational** [comment]
Yes your correct both port 25 and 587 work 

Seems I was ready some incorrect website

So all is perfect no changes needed 😊

**Comment by exileDev** [comment]
Closed #7031

**Comment by exileDev** [comment]
Closed https://github.com/nopSolutions/nopCommerce/issues/7031
